### PR TITLE
Change legend guide to object oriented approach

### PR DESCRIPTION
**Pull request by dmatos2012**
## PR Summary
Addresses #20754 to change legend guide from all `pyplot` interface to the `ax` objected oriented approach
## PR Checklist

- [N/A] Has pytest style unit tests (and `pytest` passes).
- [X] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (run `flake8` on changed files to check).
- [N/A] New features are documented, with examples if plot related.
- [X] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [N/A] Conforms to Matplotlib style conventions (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).
- [N/A] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [N/A] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).

### CODE REVIEW

File: tutorials/intermediate/legend_guide.py [hunk 1]
```diff
@@ -53,16 +53,16 @@
 For full control of what is being added to the legend, it is common to pass
 the appropriate handles directly to :func:`legend`::
 
-    line_up, = plt.plot([1, 2, 3], label='Line 2')
-    line_down, = plt.plot([3, 2, 1], label='Line 1')
-    plt.legend(handles=[line_up, line_down])
+    line_up, = ax.plot([1, 2, 3], label='Line 2')
```
Comment (story645) [inline tutorials/intermediate/legend_guide.py:56]: I think needs a 'fig, ax = plt.subplots()' to be runnable

File: tutorials/intermediate/legend_guide.py [hunk 2]
```diff
@@ -110,25 +112,25 @@
...
-#                bbox_transform=plt.gcf().transFigure)
+#     ax.legend(bbox_to_anchor=(1, 1),
+#                bbox_transform=fig.transFigure)
 #
 # More examples of custom legend placement:
 
-plt.subplot(211)
-plt.plot([1, 2, 3], label="test1")
-plt.plot([3, 2, 1], label="test2")
+ax1 = plt.subplot(211)
```
Comment (story645) [inline tutorials/intermediate/legend_guide.py:120]: fig, (ax1,ax2) = plt.subplots(nrows=2)

**Review by story645 — COMMENTED** [review]
Not sure if one of the plt examples should be left in as a resource but otherwise is awesome.

**Comment by dmatos2012** [comment]
thanks @story645  for the review. I've updated the docs acording to the reviews. However, as per issue #20553, I think it would be nice to have some sort of description of the different `loc` descriptions for axes. It can be quite tricky to  understand at first.

File: tutorials/intermediate/legend_guide.py [hunk 3]
```diff
@@ -53,16 +53,17 @@
...
+    line_down, = ax.plot([3, 2, 1], label='Line 1')
+    ax.legend(handles=[line_up, line_down])
 
 In some cases, it is not possible to set the label of the handle, so it is
 possible to pass through the list of labels to :func:`legend`::
 
-    line_up, = plt.plot([1, 2, 3], label='Line 2')
-    line_down, = plt.plot([3, 2, 1], label='Line 1')
-    plt.legend([line_up, line_down], ['Line Up', 'Line Down'])
+    line_up, = ax.plot([1, 2, 3], label='Line 2')
```
Comment (story645) [inline tutorials/intermediate/legend_guide.py:64]:
```suggestion
    fig, ax = plt.subplots()
    line_up, = ax.plot([1, 2, 3], label='Line 2')
```
yeah is repetitive but I think they're all supposed to be self contained

**Review by story645 — COMMENTED** [review]
totally agree w/ you about adding in something on location, possibly as an expansion of https://62747-1385122-gh.circle-artifacts.com/0/doc/build/html/tutorials/intermediate/legend_guide.html#legend-location & maybe using the cheatsheet as inspiration?
https://github.com/matplotlib/cheatsheets/blob/master/cheatsheets-2.png

File: tutorials/intermediate/legend_guide.py [hunk 4]
```diff
@@ -110,25 +113,24 @@
 # figure's top right-hand corner instead of the axes' corner, simply specify
 # the corner's location and the coordinate system of that location::
 #
-#     plt.legend(bbox_to_anchor=(1, 1),
-#                bbox_transform=plt.gcf().transFigure)
+#     ax.legend(bbox_to_anchor=(1, 1),
+#                bbox_transform=fig.transFigure)
```
Comment (QuLogic) [inline tutorials/intermediate/legend_guide.py:117]: Re-align.

File: tutorials/intermediate/legend_guide.py [hunk 5]
```diff
@@ -110,25 +113,24 @@
...
            ncol=2, mode="expand", borderaxespad=0.)
 
-plt.subplot(223)
-plt.plot([1, 2, 3], label="test1")
-plt.plot([3, 2, 1], label="test2")
-# Place a legend to the right of this smaller subplot.
-plt.legend(bbox_to_anchor=(1.05, 1), loc='upper left', borderaxespad=0.)
+ax2.plot([1, 2, 3], label="test1")
+ax2.plot([3, 2, 1], label="test2")
+# Place a legend on top center  of this subplot.
```
Comment (QuLogic) [inline tutorials/intermediate/legend_guide.py:132]: I think the point is to be outside on this one. We can replicate the layout with funny `subplot` mixing, using `subplot_mosaic`:
```python
fig, ax_dict = plt.subplot_mosaic([
    ['top', 'top'],
    ['bottom', '-'],
])
```

File: tutorials/intermediate/legend_guide.py [hunk 6]
```diff
@@ -144,17 +146,18 @@
...
-line2, = plt.plot([3, 2, 1], label="Line 2", linewidth=4)
+fig, ax = plt.subplots()
+line1, = ax.plot([1, 2, 3], label="Line 1", linestyle='--')
+line2, = ax.plot([3, 2, 1], label="Line 2", linewidth=4)
 
 # Create a legend for the first line.
-first_legend = plt.legend(handles=[line1], loc='upper right')
+first_legend = ax.legend(handles=[line1], loc='upper right')
 
 # Add the legend manually to the current Axes.
```
Comment (QuLogic) [inline tutorials/intermediate/legend_guide.py:156]:
```suggestion
# Add the legend manually to the Axes.
```

File: tutorials/intermediate/legend_guide.py [hunk 7]
```diff
@@ -206,24 +210,25 @@
 
 from numpy.random import randn
 
+fig, ax = plt.subplots()
 z = randn(10)
-
-red_dot, = plt.plot(z, "ro", markersize=15)
+red_dot, = ax.plot(z, "ro", markersize=15)
```
Comment (QuLogic) [inline tutorials/intermediate/legend_guide.py:215]:
```suggestion
z = randn(10)

fig, ax = plt.subplots()
red_dot, = ax.plot(z, "ro", markersize=15)
```

File: tutorials/intermediate/legend_guide.py [hunk 8]
```diff
@@ -206,24 +210,25 @@
...
 
-p1, = plt.plot([1, 2.5, 3], 'r-d')
-p2, = plt.plot([3, 2, 1], 'k-o')
+fig, ax = plt.subplots()
+p1, = ax.plot([1, 2.5, 3], 'r-d')
+p2, = ax.plot([3, 2, 1], 'k-o')
 
-l = plt.legend([(p1, p2)], ['Two keys'], numpoints=1,
+l = ax.legend([(p1, p2)], ['Two keys'], numpoints=1,
                handler_map={tuple: HandlerTuple(ndivide=None)})
```
Comment (QuLogic) [inline tutorials/intermediate/legend_guide.py:232]:
```suggestion
l = ax.legend([(p1, p2)], ['Two keys'], numpoints=1,
              handler_map={tuple: HandlerTuple(ndivide=None)})
```

File: tutorials/intermediate/legend_guide.py [hunk 9]
```diff
@@ -238,6 +243,8 @@
 
 import matplotlib.patches as mpatches
 
+fig, ax = plt.subplots()
```
Comment (QuLogic) [inline tutorials/intermediate/legend_guide.py:246]: Move down to where the legend is done.

File: tutorials/intermediate/legend_guide.py [hunk 10]
```diff
@@ -272,6 +279,8 @@ def legend_artist(self, legend, orig_handle, fontsize, handlebox):
 
 from matplotlib.legend_handler import HandlerPatch
 
+fig, ax = plt.subplots()
```
Comment (QuLogic) [inline tutorials/intermediate/legend_guide.py:282]: Move down.

File: tutorials/intermediate/legend_guide.py [hunk 11]
```diff
@@ -254,7 +261,7 @@ def legend_artist(self, legend, orig_handle, fontsize, handlebox):
         return patch
 
 
-plt.legend([AnyObject()], ['My first handler'],
+ax.legend([AnyObject()], ['My first handler'],
            handler_map={AnyObject: AnyObjectHandler()})
```
Comment (QuLogic) [inline tutorials/intermediate/legend_guide.py:265]:
```suggestion
ax.legend([AnyObject()], ['My first handler'],
          handler_map={AnyObject: AnyObjectHandler()})
```

File: tutorials/intermediate/legend_guide.py [hunk 12]
```diff
@@ -286,7 +295,7 @@ def create_artists(self, legend, orig_handle,
 
 c = mpatches.Circle((0.5, 0.5), 0.25, facecolor="green",
                     edgecolor="red", linewidth=3)
-plt.gca().add_patch(c)
+ax.add_patch(c)
 
-plt.legend([c], ["An ellipse, not a rectangle"],
+ax.legend([c], ["An ellipse, not a rectangle"],
            handler_map={mpatches.Circle: HandlerEllipse()})
```
Comment (QuLogic) [inline tutorials/intermediate/legend_guide.py:301]:
```suggestion
ax.legend([c], ["An ellipse, not a rectangle"],
          handler_map={mpatches.Circle: HandlerEllipse()})
```

File: tutorials/intermediate/legend_guide.py (hunk 5 above)
Comment (dmatos2012) [reply tutorials/intermediate/legend_guide.py]: This is actually pretty cool, wasnt aware of that, bc I was struggling with using `fig, ax = plt.subplots()` to get that layout, hence changing the label to the inside. Thx for the tip @QuLogic

**Comment by dmatos2012** [comment]
> totally agree w/ you about adding in something on location, possibly as an expansion of https://62747-1385122-gh.circle-artifacts.com/0/doc/build/html/tutorials/intermediate/legend_guide.html#legend-location & maybe using the cheatsheet as inspiration?
> https://github.com/matplotlib/cheatsheets/blob/master/cheatsheets-2.png

I was thinking perhaps something like creating a 3x3 (or 2 x2 w.e) `subplot_mosaic` and just placing the legend in the different `loc` positions, and then link to the `cheatsheet.png` for their future reference. But given that this is mentioned on a different issue, I was thinking on creating another PR just for that, unless you clearly want it on this one instead. lmk

**Comment by story645** [comment]
I think it's a good idea to leave it to another issue since that's gonna be a bigger change to the guide content so folks might have more opinions.

File: tutorials/intermediate/legend_guide.py [hunk 13]
```diff
@@ -205,26 +211,27 @@
 # combining two legend keys on top of one another:
 
 from numpy.random import randn
-
```
Comment (QuLogic) [inline tutorials/intermediate/legend_guide.py:208]: Keep blank line after import.

Comment (dmatos2012) [reply tutorials/intermediate/legend_guide.py:208]: done :)

**Review by story645 — APPROVED** [review]

**Pull request merged**